### ISSUE TITLE
Contributing

### ISSUE THREAD

**Issue by elalaouifaris** [queryverse/QuickVega.jl#4]
Hi,

I just saw the Juliacon video on the Vega ecosystem. I guess this package is a good way to start contributing?
What should I check in the Queryverse ecosystem to start working on this?
I'm a newbie in Julia development, but I worked with the Tidyverse in the R ecosystem. 
Switching to Julia now! 

Thanks
Faris

**Comment by davidanthoff** [comment]
Yes, that would be awesome!

My general idea for this package is that it just exports a bunch of simple functions that all return either a Vega.jl or VegaLite.jl spec. The idea generally is that the API _in this package here_ is not grammar of graphics style, but just one-function-one-plot, super simple.

The prototype of that would probably be something like a `histogram` function:
```julia
function histogram(data)
  return @vlplot(:bar, x={data, bin=true}, y="count()")
end
```
or something like that. It doesn't save you a lot of code, but I for example can never remember the details of this, so just having a short and sweet function for something as common like this seems really helpful.

Broadly speaking, I think there are two categories of potential functions:

1) Simple functions for very simple plots that are currently a bit too verbose when you create them with `@vlplot`. The `histogram` case is an example of this. I should say, that for some of these my hope is to eventually make these cases simpler in the VegaLite.jl repo itself, or in vega-lite, so if we add some of these here it might be more a short to medium term fix. To stick with the histogram example: I think the vega-lite team might add a new mark `histogram` at some point, and then creating a histogram with the `@vlplot` macro would look like `@vlplot(:histogram, data)`. I think at that point that is concise enough that we wouldn't need a `histogram` function in this package here anymore. But, that is future talk, and I think in the meantime some of these shorter functions would really help.

2) really complicated plots that one can create with one simple call. Seaborn has a lot of examples for that. Take for example the pair plot from [here](https://seaborn.pydata.org/generated/seaborn.pairplot.html#seaborn.pairplot). I don't think creating something like that will ever be concise in original VegaLite.jl, so having more like "shortcut" functions here that do this would be great.

There was also an older version of the Vega.jl package that had an API that was more in this spirit, the documentation is still up [here](http://johnmyleswhite.github.io/Vega.jl/), and there might be ideas there as well.

I think one nice thing about this package is that one can get started very piecemeal, by just adding say one new function. No need to integrate with some pre-existing complicated architecture, every function here on its own helps :)

**Comment by hsm207** [comment]
@davidanthoff Is it okay to dump all functions in QuickVega.jl for now or is there already something like a design about how the source code should look like?

**Comment by elalaouifaris** [comment]
@davidanthoff Thank you for the info.
Maybe we can start by adding a few issues to document with few example functions and gradually launch pull requests for them?
It's a really good case for first contributions, it can give the opportunity for new contributors (like me) to get the feel for the github workflow.

**Comment by davidanthoff** [comment]
I think we could just start by dumping things into the repo and then discuss based on that? I really don't have any strong ideas for this beyond what I wrote above, so I think this would be very much a collaborative trial and error process for those that want to participate :) 

I think maybe a good strategy might also be to say that QuickVega.jl is in experimental state for the next six months or something like that, and during that time we might at any time change everything, i.e. we would provide zero backwards compat promises. That might allow us to really experiment without complicated backwards compat constraints.

So I think just PRs like https://github.com/queryverse/QuickVega.jl/pull/5 are great: we can just start, see what works and take it from there.

Really excited to see that you all are interested in contributing!!

**Comment by davibarreira** [comment]
Hey guys, are you still working on this package? I didn't know that this initiative existed, so I was starting to implement it myself, with a package called [Aquarius.jl](https://github.com/davibarreira/Aquarius.jl).

**Comment by hsm207** [comment]
@davibarreira I'm still interested in contributing but I just don't have any idea what visualizations to start with. If you have ideas, I'd be happy to help implement them here.

**Comment by davibarreira** [comment]
I created a README.md file and a logo for the project (personal hobby). I also added myself in the Project.toml (never contributed to open source, so I'm sorry if I wasn't supposed to do that).

**Comment by davibarreira** [comment]
So, I forked the repo, added a little README and created a logo (personal hobby). Since this package was created before mine, I guess it is best to work on it instead of creating another./

**Comment by davibarreira** [comment]
@hsm207 , I have already quite some ideas in mind. But the pull requests submitted by @elalaouifaris are still pending since August. So I'm wondering if this package is still active. If not, perhaps we can start a new one. As I said, I already started Aquarius.jl.

**Comment by hsm207** [comment]
@davibarreira I'd consider this package as active since davidanthoff is still around in the Julia community. He's probably busy with teaching and/or other projects.

Why don't we work on a fork of this project instead? I suggest this because I think it benefits this project in terms of discoverability to be part of the queryverse organization.

**Comment by davibarreira** [comment]
Yeah! David actually responded to me on Slack. The package is active, which is great. Working on the fork seems like a good alternative.

**Comment by hsm207** [comment]
@davibarreira I see you've already forked the project. I have a local copy of that fork.

What should we get started with?

**Comment by davibarreira** [comment]
@hsm207 , I haven't planned much,  but at the moment I guess we could start by implementing some of the basic plots, such as 2 of the pull requests have already done. For example, there is already an implementation of histogram and lineplot. Each implementation has a version using the data as DataFrame, and another using Arrays, which allows the package to be independent of DataFrames.jl. So I guess we could continue on this line, implementing some other basic plots, such as barplots, trailplots, distplots, etc.
There are many other things to think about, such as how to allow kwargs to functions, so we can pass attributes such as color, field types, markers, labels, etc. Later I will try to write down the list of things that I think would be cool to have in the package, and I'll share it in my fork, so we can have some sort of road map to start with.

**Comment by davibarreira** [comment]
So, I'm implementing some functions and I'm facing a dilemma in terms of design. For functions such as "barplot", "lineplot", "scatterplot", would it be better to have a unique plotting function, like "plot", that accepts different marks (line, bar, circle), or to do like Seaborn, and do an implementation for each? I've started with a function for each, but the code is looking quite redundant.

**Comment by davibarreira** [comment]
Hey @davidanthoff , I've finally restarted working on QuickVega, and will shortly send a pull request with several contributions. I know that you didn't want to make the package dependent on DataFrames.jl, but I've encountered some issues (I'll clarify in the pull request), so for now some of my code does depend on DataFrames. I've decided to keep on going like this for now, but the code can be easily refactored later to be made independent of DataFrames. Although, now that DataFrames is in version 1.0, perhaps it's more stable and can become a dependence. What do you think? Cheers!

**Comment by davidanthoff** [comment]
Great to hear this and looking forward to the PR!

I think at this stage I'm primarily excited about any contribution to this package, and we can figure things like DataFrames.jl dependencies out as we go along.

**Comment by davibarreira** [comment]
The code is still very undocumented, so I'll be doing some due diligence, and then submit the pull request.

**Comment by davibarreira** [comment]
@hsm207 , you were looking to contribute earlier, so you might want to take a look at the pull request I just submitted. I think I made some good advances, specially on that issue of the `kwargs`. If David accepts the pull request, I think the package can advance quite fast from here on out.
